File: api/v1/endpoints/concept_themes.py
# -*- coding: utf-8 -*-
"""Concept/theme graph, membership, exposure, beta, and alpha endpoints."""

from fastapi import APIRouter, HTTPException, Query

from src.services.concept_theme_service import ConceptThemeError, ConceptThemeService
from src.services.concept_theme_worker import ConceptThemeWorker

router = APIRouter()


def _http_error(exc: Exception, status: int = 422) -> HTTPException:
    return HTTPException(status_code=status, detail={"error": "concept_theme_error", "message": str(exc)[:500]})


@router.get("/overview", summary="读取分层概念题材总览")
def overview(
    query: str = Query(default="", max_length=100),
    theme_type: str = Query(default="", max_length=24),
    source: str = Query(default="", max_length=24),
    family: str = Query(default="", max_length=80),
    cluster: str = Query(default="", max_length=80),
    sort_by: str = Query(default="heat", pattern="^(heat|name|size|change)$"),
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=80, ge=12, le=200),
):
    return ConceptThemeService().overview(
        query=query, theme_type=theme_type, source=source, family=family, cluster=cluster,
        sort_by=sort_by, page=page, page_size=page_size,
    )


@router.get("/themes/{theme_id}", summary="读取题材成分、权重及Beta/Alpha")
def theme_detail(
    theme_id: int,
    refresh_if_empty: bool = Query(default=True),
    horizon_days: int = Query(default=60, ge=20, le=120),
):
    try:
        return ConceptThemeService().theme_detail(
            theme_id, refresh_if_empty=refresh_if_empty, horizon_days=horizon_days,
        )
    except ConceptThemeError as exc:
        raise _http_error(exc, 404 if "不存在" in str(exc) else 422)


@router.get("/stocks/{ts_code}", summary="读取个股题材暴露与独特Alpha证据")
def stock_lens(
    ts_code: str,
    refresh_if_empty: bool = Query(default=True),
    horizon_days: int = Query(default=60, ge=20, le=120),
):
    try:
        return ConceptThemeService().stock_lens(
            ts_code, refresh_if_empty=refresh_if_empty, horizon_days=horizon_days,
        )
    except ConceptThemeError as exc:
        raise _http_error(exc)


@router.get("/status", summary="读取题材库同步状态")
def status():
    return {**ConceptThemeService().sync_status(), "worker": ConceptThemeWorker.get_instance().status()}


@router.get("/rotation", summary="读取多源题材轮动强弱")
def rotation(
    days: int = Query(default=20, ge=5, le=60),
    limit: int = Query(default=24, ge=8, le=60),
):
    return ConceptThemeService().rotation(days=days, limit=limit)


@router.post("/sync/catalog", status_code=202, summary="唤醒题材库后台更新")
def wake_catalog_sync():
    return ConceptThemeWorker.get_instance().trigger()


@router.post("/themes/{theme_id}/refresh", status_code=202, summary="直接更新单一题材及归因")
def refresh_theme(theme_id: int):
    try:
        return ConceptThemeService().refresh_theme(theme_id, calculate=True)
    except ConceptThemeError as exc:
        raise _http_error(exc)


@router.post("/stocks/{ts_code}/refresh", status_code=202, summary="直接更新单股题材与归因")
def refresh_stock(ts_code: str):
    try:
        return ConceptThemeService().refresh_stock(ts_code, calculate=True)
    except ConceptThemeError as exc:
        raise _http_error(exc)

File: apps/dsa-web/src/api/conceptThemes.ts
import apiClient, { BACKGROUND_ROUTE_HEADERS } from './index';
import { cachedQuery, invalidateCachedQueries } from './requestCache';
import { toCamelCase } from './utils';

export type ConceptTheme = {
  id: number;
  source: string;
  sourceLabel: string;
  sourceCode: string;
  name: string;
  canonicalName: string;
  themeType: string;
  level: number;
  parentCode?: string | null;
  constituentCount: number;
  marketDate?: string | null;
  heatScore?: number | null;
  pctChange?: number | null;
  fundFlow?: number | null;
  family: string;
  cluster: string;
  updatedAt?: string | null;
};

export type ConceptStock = {
  tsCode: string;
  name: string;
  sources: string[];
  reasons: string[];
  sourceCount: number;
  weightScore: number;
  beta?: number | null;
  marketBeta?: number | null;
  alphaAnnualized?: number | null;
  residualReturn?: number | null;
  rSquared?: number | null;
  observations?: number;
  confidence: string;
  betaInterpretation?: string;
  components?: Record<string, number | string | null>;
};

export type ConceptOverview = {
  items: ConceptTheme[];
  total: number;
  page: number;
  pageSize: number;
  summary: {
    themes: number;
    memberships: number;
    memberedThemes: number;
    attemptedThemes: number;
    failedThemes: number;
    scanCoveragePct: number;
    membershipCoveragePct: number;
    exposures: number;
    sources: Record<string, number>;
    types: Record<string, number>;
    families: Record<string, number>;
    clusterFamilies: Record<string, Record<string, number>>;
    marketDate?: string;
  };
  sync?: { status: string; progress: number; stage: string; marketDate?: string; sources?: Record<string, number> } | null;
  methodology: ConceptMethodology;
};

export type ConceptMethodology = {
  version: string;
  principles: string[];
  weightFormula: Record<string, number>;
  betaFormula: string;
  windows: number[];
  minimumObservations: number;
  sources: Array<{ key: string; name: string; reliability: number }>;
  licenseNote: string;
};

export type ThemeDetail = {
  theme: ConceptTheme;
  sourceNodes: ConceptTheme[];
  stocks: ConceptStock[];
  totalStocks: number;
  consensusStocks: number;
  consensusDistribution?: { strong: number; confirmed: number; singleSource: number };
  attributionReady: number;
  horizonDays: number;
  methodology: ConceptMethodology;
};

export type StockThemeLens = {
  tsCode: string;
  name: string;
  asOfDate?: string | null;
  horizonDays: number;
  themes: Array<ConceptStock & {
    canonicalName: string;
    family: string;
    cluster: string;
    themeType: string;
    themeIds: number[];
  }>;
  primaryThemes: StockThemeLens['themes'];
  uniqueDrivers: Array<{
    kind: string;
    title: string;
    summary?: string;
    source: string;
    date?: string;
    importance?: number;
    url?: string;
  }>;
  summary: { themeCount: number; sourceCount: number; consensusCount: number; alphaPositiveCount: number };
  methodology: ConceptMethodology;
};

type OverviewParams = {
  query?: string;
  themeType?: string;
  source?: string;
  family?: string;
  cluster?: string;
  sortBy?: 'heat' | 'name' | 'size' | 'change';
  page?: number;
  pageSize?: number;
};

type RawConceptOverview = {
  summary?: {
    sources?: Record<string, number>;
    types?: Record<string, number>;
    families?: Record<string, number>;
    cluster_families?: Record<string, Record<string, number>>;
  };
  sync?: { sources?: Record<string, number> } | null;
};

/**
 * API 字段名需要转成 camelCase，但题材名和来源代码本身是业务数据，不能改写。
 * camelcase-keys 的 deep 模式会把 “AI算力…” 误改为 “ai算力…”，继而让筛选条件失配。
 */
export function normalizeConceptOverview(data: unknown): ConceptOverview {
  const normalized = toCamelCase<ConceptOverview>(data);
  const raw = data as RawConceptOverview;
  if (raw.summary) {
    normalized.summary.sources = raw.summary.sources ?? {};
    normalized.summary.types = raw.summary.types ?? {};
    normalized.summary.families = raw.summary.families ?? {};
    normalized.summary.clusterFamilies = raw.summary.cluster_families ?? {};
  }
  if (normalized.sync && raw.sync?.sources) normalized.sync.sources = raw.sync.sources;
  return normalized;
}

export const conceptThemesApi = {
  overview: async (params: OverviewParams = {}): Promise<ConceptOverview> => {
    const key = `concept:overview:${JSON.stringify(params)}`;
    return cachedQuery(key, async () => {
      const response = await apiClient.get('/api/v1/concept-themes/overview', {
        params: {
          query: params.query || undefined,
          theme_type: params.themeType || undefined,
          source: params.source || undefined,
          family: params.family || undefined,
          cluster: params.cluster || undefined,
          sort_by: params.sortBy ?? 'heat',
          page: params.page ?? 1,
          page_size: params.pageSize ?? 80,
        },
        headers: BACKGROUND_ROUTE_HEADERS,
      });
      return normalizeConceptOverview(response.data);
    }, { freshMs: 20_000, staleMs: 5 * 60_000 });
  },
  theme: async (themeId: number, refreshIfEmpty = true, horizonDays = 60): Promise<ThemeDetail> => {
    const response = await apiClient.get(`/api/v1/concept-themes/themes/${themeId}`, {
      params: { refresh_if_empty: refreshIfEmpty, horizon_days: horizonDays },
      headers: BACKGROUND_ROUTE_HEADERS,
      timeout: 90_000,
    });
    return toCamelCase<ThemeDetail>(response.data);
  },
  stock: async (tsCode: string, refreshIfEmpty = true, horizonDays = 60): Promise<StockThemeLens> => {
    const response = await apiClient.get(`/api/v1/concept-themes/stocks/${encodeURIComponent(tsCode)}`, {
      params: { refresh_if_empty: refreshIfEmpty, horizon_days: horizonDays },
      headers: BACKGROUND_ROUTE_HEADERS,
      timeout: 90_000,
    });
    return toCamelCase<StockThemeLens>(response.data);
  },
  wakeSync: async (): Promise<void> => {
    await apiClient.post('/api/v1/concept-themes/sync/catalog', {});
    invalidateCachedQueries('concept:');
  },
};
